### PR TITLE
Treeview: Add symbol tree

### PR DESCRIPTION
**Pull request by philipp-spiess**
Part of #12916 

This adds a new implementation of the wildcard `<Tree>` component to render symbols. The react component we use for the Tree already provides a nice abstraction to convert the exact same object structure that we compute into the internal representation needed for the tree view. I've forked this helper function to make it a bit more generic so it keeps arbitrary node data alive.

A bit of extra book keeping is necessary to only update the selected IDs when necessary but it was possible to simplify this a lot by throwing away the component state whenever the query data changes (at which point the treeview, which relies on index offsets, was very unhappy. It's only save to append to this data structure).

## Test plan

- Added behind a feature flag
- Lots of manual testing
- Dogfooding!

https://user-images.githubusercontent.com/458591/219678946-0f6038ea-a093-454f-8ab7-7040b20023ae.mov

## App preview:

- [Web](https://sg-web-ps-symbol-tree.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: client/web/src/repo/RepoRevisionSidebarSymbolTree.tsx [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    }, [] as number[])
+
+    // const lastUrlNavigatedToRef = React.useRef<null | string>(null)
+    const onSelect = useCallback(
+        ({ element, isSelected }: { element: SymbolNode; isSelected: boolean }) => {
+            if (!isSelected || element.__typename !== 'Symbol' || selectedSymbolUrl === element.url) {
+                return
+            }
+
+            // lastUrlNavigatedToRef.current = element.url
```
Comment (taras-yemets) [inline client/web/src/repo/RepoRevisionSidebarSymbolTree.tsx:59]:
```suggestion
```

File: client/web/src/repo/RepoRevisionSidebarSymbolTree.tsx [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    // Mark all expandable nodes as expanded by default.
+    const defaultExpandedIds = treeData.reduce((acc, node) => {
+        if (node.children.length > 0) {
+            acc.push(node.id)
+        }
+        return acc
+    }, [] as number[])
+
+    // const lastUrlNavigatedToRef = React.useRef<null | string>(null)
```
Comment (taras-yemets) [inline client/web/src/repo/RepoRevisionSidebarSymbolTree.tsx:52]:
```suggestion
```

File: client/wildcard/src/components/Tree/react-accessible-treeview/utils.ts [hunk 3]
```diff
@@ -214,6 +214,7 @@ export const flattenTree = function (tree: ITreeNode): INode[] {
     const flattenedTree: INode[] = []
 
     const flattenTreeHelper = function (tree: ITreeNode, parent: number | null) {
+        console.log({ tree, parent })
```
Comment (taras-yemets) [inline client/wildcard/src/components/Tree/react-accessible-treeview/utils.ts:217]:
```suggestion
```

**Review by taras-yemets — APPROVED** [review]
Great work, @philipp-spiess! 
The code looks good to me.
Checked out the branch locally. It feels like a huge UX improvement compared to the existing symbols panel.

File: client/web/src/repo/RepoRevisionSidebarSymbolTree.tsx (hunk 1 above)
Comment (philipp-spiess) [reply client/web/src/repo/RepoRevisionSidebarSymbolTree.tsx:59]: whoops

**Review by limitedmage — APPROVED** [review]
Very cool!

Any chance we can add some tests, eg for `flattenTree`?

**Comment by philipp-spiess** [comment]
> Any chance we can add some tests, eg for flattenTree?

I'll also be looking at migrating the existing tree tests in a follow-up PR!

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (+0.04 kb) | 0.02% (+2.33 kb) | 0.02% (+2.29 kb) | 0.13% (+1) 🔺 |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/ps/symbol-tree/compare-report.html;tab=live_object) for a full comparison between the commits e079f15 and 37458d2 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Pull request merged**